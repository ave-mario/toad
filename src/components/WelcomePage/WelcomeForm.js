import React from 'react';
import { withFormik } from 'formik';
import { string, object } from 'yup';
import PropTypes from 'prop-types';
import { Header, Main, Footer, ErrorText } from '../../elements';
import { dispatcher } from '../../config/redux.store';
import authActions from '../../actions/auth.actions';
import reg from '../../config/regex';
import {
  SubmitButton,
  WelcomeForm,
  WelcomeInput,
  WelcomeTitle,
  WelcomeSubTitle
} from './elements/welcome.form';

const { Creators } = authActions;

export const validationSchema = object().shape({
  password: string()
    .required('password is required')
    .matches(reg.password, 'The password cannot contain spaces')
});

export const handleSubmit = (values, { props }) => {
  const { token } = props;
  dispatcher(Creators.createPasswordRequest(values.password, token));
};

export const mapPropsToValues = () => ({
  password: ''
});

export const MyFormInner = props => {
  const {
    values,
    handleChange,
    handleBlur,
    errors,
    touched,
    handleSubmit: submit,
    isSubmitting,
    isFailed,
    name
  } = props;
  return (
    <WelcomeForm onSubmit={submit}>
      <Header>
        <WelcomeTitle>Welcome{name && `, ${name}`}</WelcomeTitle>
        <WelcomeSubTitle>Please, create a password</WelcomeSubTitle>
      </Header>
      <Main>
        <WelcomeInput
          className="passwordInput"
          type="password"
          name="password"
          placeholder="Password"
          onChange={handleChange}
          border={errors.password && touched.password && '1px solid red'}
          value={values.password}
          onBlur={handleBlur}
        />
        {errors.password && touched.password && (
          <ErrorText>{errors.password}</ErrorText>
        )}
      </Main>
      <Footer>
        <SubmitButton
          className="submitButton"
          type="submit"
          disabled={isSubmitting && !isFailed}
        >
          CREATE
        </SubmitButton>
      </Footer>
    </WelcomeForm>
  );
};

MyFormInner.propTypes = {
  values: PropTypes.objectOf(PropTypes.string).isRequired,
  handleChange: PropTypes.func.isRequired,
  handleBlur: PropTypes.func.isRequired,
  errors: PropTypes.objectOf(PropTypes.string).isRequired,
  touched: PropTypes.objectOf(PropTypes.bool).isRequired,
  handleSubmit: PropTypes.func.isRequired,
  isSubmitting: PropTypes.bool.isRequired,
  isFailed: PropTypes.bool,
  name: PropTypes.string
};
MyFormInner.defaultProps = {
  isFailed: false,
  name: ''
};

export const Welcome = withFormik({
  mapPropsToValues,
  validationSchema,
  handleSubmit
})(MyFormInner);
