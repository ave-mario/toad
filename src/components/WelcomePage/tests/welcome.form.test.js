import React from 'react';
import { shallow, mount } from 'enzyme';
import { Welcome } from '../WelcomeForm';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

describe('Welcome form', () => {
  const mockStore = configureStore();
  const store = mockStore({
    auth: {
      isFailed: false,
      user: {},
      error: {
        message: 'message',
        success: false
      },
      createPassword: {
        success: null,
        error: null
      }
    }
  });

  const wrapper = mount(
    <Provider store={store}>
      <Welcome />
    </Provider>
  );
  test('should update an password input when it is changed', () => {
    wrapper.find('input.passwordInput').simulate('change', {
      persist: () => {},
      target: {
        name: 'password',
        value: 'abc'
      }
    });
    wrapper.update();
    let myFormInner = wrapper.find('MyFormInner');
    expect(myFormInner.props().values.password).toBe('abc');
    wrapper.update();
    myFormInner = wrapper.find('MyFormInner');
  });
  test('submit', () => {
    let prevented = false;
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {
        prevented = true;
      }
    });
    expect(prevented).toBe(true);
  });
});

describe('Welcome form snaps', () => {
  test('should render login form', () => {
    const component = shallow(<Welcome />);

    expect(component).toMatchSnapshot();
  });
});
