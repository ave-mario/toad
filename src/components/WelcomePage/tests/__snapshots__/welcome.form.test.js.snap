// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Welcome form snaps should render login form 1`] = `
<Formik
  enableReinitialize={false}
  handleSubmit={[Function]}
  initialValues={
    Object {
      "password": "",
    }
  }
  isInitialValid={false}
  onSubmit={[Function]}
  render={[Function]}
  validateOnBlur={true}
  validateOnChange={true}
  validationSchema={[Function]}
/>
`;
